lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1898,6 +1898,49 @@ LogicalResult AttachOp::canonicalize(AttachOp op, PatternRewriter &rewriter) {
   return failure();
 }
 
+/// Replaces the given op with the contents of the given single-block region.
+static void replaceOpWithRegion(PatternRewriter &rewriter, Operation *op,
+                                Region &region) {
+  assert(llvm::hasSingleElement(region) && "expected single-region block");
+  Block *fromBlock = &region.front();
+  // Merge it in above the specified operation.
+  op->getBlock()->getOperations().splice(Block::iterator(op),
+                                         fromBlock->getOperations());
+}
+
+LogicalResult WhenOp::canonicalize(WhenOp op, PatternRewriter &rewriter) {
+  if (auto constant = op.getCondition().getDefiningOp<firrtl::ConstantOp>()) {
+    if (constant.getValue().isAllOnes())
+      replaceOpWithRegion(rewriter, op, op.getThenRegion());
+    else if (!op.getElseRegion().empty())
+      replaceOpWithRegion(rewriter, op, op.getElseRegion());
+
+    rewriter.eraseOp(op);
+
+    return success();
+  }
+
+  // Erase empty if-else block.
```

Comment:
This feels like this could be simplified (here to end) but no concrete suggestion.

I think this looks right, it takes care to not bother deleting the else region if the then region is also empty (preferring to directly drop the entire op).